### PR TITLE
682.hwi.improve wrong login response

### PR DESCRIPTION
**Pull request by gbasolu**
There is a parallel MR in HWI

### CODE REVIEW

File: hydra_base/lib/users.py [hunk 1]
```diff
@@ -22,10 +22,11 @@
 
 from .objects import JSONObject
 from ..db.model import User, Role, Perm, RoleUser, RolePerm
-from ..exceptions import ResourceNotFoundError, HydraError
+from ..exceptions import ResourceNotFoundError, HydraError, HydraLoginUserNotFound, HydraLoginUserMaxAttemptsExceeded, HydraLoginUserPasswordWrong
 from .. import db
 from .. import config
 from ..util.permissions import required_perms
+import transaction
```
Comment (knoxsp) [inline hydra_base/lib/users.py:29]: Hydra base does not use transactions in this way. Transactions should be managed using the 'commit_transaction' function in db/__init__.py. 

We never commit transactions inside functions in this way.

File: hydra_base/lib/users.py [hunk 2]
```diff
@@ -248,19 +249,21 @@ def inc_failed_login_attempts(username, flush=True, **kwargs):
...
     if user_i.failed_logins is not None:
         user_i.failed_logins = user_i.failed_logins + 1
     else:
         user_i.failed_logins = 1
 
+    log.info("[ inc_failed_login_attempts ] User {} has currently {} failed logins".format(username, user_i.failed_logins))
     if flush is True:
         db.DBSession.flush()
+        # This is needed
+        transaction.commit()
```
Comment (knoxsp) [inline hydra_base/lib/users.py:266]: See my previous comment. Do not commit this transaction. If it is needed, then it needs to be solved elsewhere. Why is it needed exactly?

File: hydra_base/lib/users.py [hunk 3]
```diff
@@ -281,12 +284,14 @@ def reset_failed_logins(username, flush=True, **kwargs):
...
     except NoResultFound:
-        raise HydraError(username)
+        # raise HydraError(username)
+        raise HydraLoginUserNotFound("User '{}' does not exist".format(username))
 
     user_i.failed_logins = 0
 
     if flush is True:
         db.DBSession.flush()
+        transaction.commit()
```
Comment (knoxsp) [inline hydra_base/lib/users.py:294]: See previous comments.

**Review by knoxsp — COMMENTED** [review]
Are there unit tests to check these different login scenarios? If not, please add them.

File: hydra_base/lib/users.py (hunk 2 above)
Comment (gbasolu) [reply hydra_base/lib/users.py:266]: Without the commit, a following flush got locked forever and breaks the code flow with a db exception on the max-lock-timeout exceeded.

File: hydra_base/lib/users.py (hunk 3 above)
Comment (gbasolu) [reply hydra_base/lib/users.py:294]: Without the commit, a following flush got locked forever and breaks the code flow with a db exception on the max-lock-timeout exceeded.

File: hydra_base/lib/users.py (hunk 1 above)
Comment (gbasolu) [reply hydra_base/lib/users.py:29]: I found in make_user_root so I used it. 
But I am going to use commit_transaction

File: hydra_base/lib/users.py (hunk 3 above)
Comment (knoxsp) [reply hydra_base/lib/users.py:294]: Hmmm. I wonder is this then not a part of the scoped session workflow? This should be taken care of by zope.
Ah wait. Yes, i see. Question is: where is the this login called from? HWI?

File: hydra_base/lib/users.py (hunk 1 above)
Comment (knoxsp) [reply hydra_base/lib/users.py:29]: Ok, Please remove it from there also.

File: hydra_base/lib/users.py (hunk 3 above)
Comment (gbasolu) [reply hydra_base/lib/users.py:294]: Yes it is called from HWI

File: hydra_base/lib/users.py (hunk 1 above)
Comment (gbasolu) [reply hydra_base/lib/users.py]: Done. What are the valid test? Tests or unittests?

File: tests/test_login.py [hunk 4]
```diff
@@ -68,3 +68,35 @@ def test_get_session_user(self, client):
         retrieved_user_id = client.get_session_user(session_id=session_id)
 
         assert retrieved_user_id == user_id
+
+    def test_login_wrong_user(self, client):
+        try:
```
Comment (knoxsp) [inline tests/test_login.py:73]: This does not conform to the standard we use for checking exceptions in the hydra-base tests. 
We use the following format:
```python
with pytest.raises(hydra_base.exceptions.HydraLoginUserNotFound):
    user_id, session_id = client.login('wrong-user!', '')
```

Comment (knoxsp) [reply tests/test_login.py:73]: Please look at the other tests for reference

File: tests/test_login.py [hunk 5]
```diff
@@ -68,3 +68,35 @@ def test_get_session_user(self, client):
...
+            user_id, session_id = client.login('wrong-user!', '')
+            assert False
+        except hydra_base.exceptions.HydraLoginUserNotFound as e:
+            assert True
+        except:
+            assert False
+
+
+    def test_login_wrong_password(self, client):
+        try:
```
Comment (knoxsp) [inline tests/test_login.py:83]: as above, please use the standard way of testing for exceptions

**Review by knoxsp — CHANGES_REQUESTED** [review]

File: tests/test_login.py (hunk 4 above)
Comment (gbasolu) [reply tests/test_login.py:73]: Done it

File: tests/test_login.py (hunk 5 above)
Comment (gbasolu) [reply tests/test_login.py:83]: done it

File: hydra_base/lib/users.py (hunk 3 above)
Comment (knoxsp) [reply hydra_base/lib/users.py]: has this been removed?

File: hydra_base/lib/users.py (hunk 2 above)
Comment (knoxsp) [reply hydra_base/lib/users.py]: has this been removed?

**Review by knoxsp — APPROVED** [review]

**Pull request merged**